### PR TITLE
chore: add static typing with mypy

### PR DESCRIPTION
**Pull request by fpgmaas**
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

This adds static typing with `mypy`. This makes it more clear to end users what expected arguments are, and helps us catch bugs earlier.

Based on top of https://github.com/MrPowers/chispa/pull/129, so this needs to be rebased and reviewd after that one is merged.

### CODE REVIEW

File: chispa/__init__.py [hunk 1]
```diff
@@ -46,28 +49,26 @@
 
 
 class Chispa:
-    def __init__(self, formats: FormattingConfig | None = None, default_output=None):
+    def __init__(self, formats: FormattingConfig | None = None) -> None:
```
Comment (SemyonSinchenko) [inline chispa/__init__.py:52]: Is it a breaking change? Why did we drop `default_output`?

File: chispa/bcolors.py [hunk 2]
```diff
@@ -1,43 +0,0 @@
...
-    # Style
-    Bold = "\033[1m"
-    Underline = "\033[4m"
-
-
-def blue(s: str) -> str:
-    return bcolors.LightBlue + str(s) + bcolors.LightRed
-
-
-def underline_text(input_text: str) -> str:
```
Comment (SemyonSinchenko) [inline chispa/bcolors.py:39]: It is a [part](https://mrpowers.github.io/chispa/reference/chispa/bcolors/) of the public API. I do not like the idea of dropping it without even deprecation warning.

File: chispa/column_comparer.py [hunk 3]
```diff
@@ -11,35 +12,33 @@ class ColumnsNotEqualError(Exception):
     pass
 
 
-def assert_column_equality(df, col_name1, col_name2):
-    elements = df.select(col_name1, col_name2).collect()
-    colName1Elements = list(map(lambda x: x[0], elements))
-    colName2Elements = list(map(lambda x: x[1], elements))
+def assert_column_equality(df: DataFrame, col_name1: str, col_name2: str) -> None:
+    rows = df.select(col_name1, col_name2).collect()
+    colName1Elements = list(map(lambda x: x[0], rows))
```
Comment (SemyonSinchenko) [inline chispa/column_comparer.py:17]: If we started to do small changes here, may we rename it to snake case? I would also replace `list(map(lambda))` by list comprehension as a more readable.

File: chispa/column_comparer.py [hunk 4]
```diff
@@ -11,35 +12,33 @@
...
         raise ColumnsNotEqualError("\n" + t.get_string())
 
 
-def assert_approx_column_equality(df, col_name1, col_name2, precision):
-    elements = df.select(col_name1, col_name2).collect()
-    colName1Elements = list(map(lambda x: x[0], elements))
-    colName2Elements = list(map(lambda x: x[1], elements))
+def assert_approx_column_equality(df: DataFrame, col_name1: str, col_name2: str, precision: float) -> None:
+    rows = df.select(col_name1, col_name2).collect()
+    colName1Elements = list(map(lambda x: x[0], rows))
```
Comment (SemyonSinchenko) [inline chispa/column_comparer.py:32]: Let's change to snake case and list comprehension maybe?

File: chispa/number_helpers.py [hunk 5]
```diff
@@ -1,18 +1,19 @@
...
-def isnan(x):
+def isnan(x: Any) -> bool:
     try:
         return math.isnan(x)
     except TypeError:
         return False
 
 
-def nan_safe_equality(x, y) -> bool:
+def nan_safe_equality(x: int | float, y: int | float) -> bool:
```
Comment (SemyonSinchenko) [inline chispa/number_helpers.py:14]: `decimal.Decimal` is a valid input too. Using decimals is very common in Spark.

File: chispa/number_helpers.py [hunk 6]
```diff
@@ -1,18 +1,19 @@
...
     return (x == y) or (isnan(x) and isnan(y))
 
 
-def nan_safe_approx_equality(x, y, precision) -> bool:
+def nan_safe_approx_equality(x: int | float, y: int | float, precision: float) -> bool:
```
Comment (SemyonSinchenko) [inline chispa/number_helpers.py:18]: `Decimal` should be a valid input too

File: chispa/schema_comparer.py [hunk 7]
```diff
@@ -44,33 +50,33 @@ def inner(s1, s2, ignore_nullable, ignore_metadata):
 # deprecate this
```
Comment (SemyonSinchenko) [inline chispa/schema_comparer.py:44]: Let's, maybe, add a deprecation warning?

File: chispa/schema_comparer.py [hunk 8]
```diff
@@ -44,33 +50,33 @@
...
         for sf1, sf2 in zipped:
             if sf1 == sf2:
-                t.add_row([blue(sf1), blue(sf2)])
+                t.add_row([blue(str(sf1)), blue(str(sf2))])
             else:
                 t.add_row([sf1, sf2])
         raise SchemasNotEqualError("\n" + t.get_string())
 
 
 # deprecate this.  ignore_nullable should be a flag.
```
Comment (SemyonSinchenko) [inline chispa/schema_comparer.py:59]: Let's, maybe, add a deprecation warning?

File: chispa/schema_comparer.py [hunk 9]
```diff
@@ -44,33 +50,33 @@
...
         for sf1, sf2 in zipped:
             if are_structfields_equal(sf1, sf2, True):
-                t.add_row([blue(sf1), blue(sf2)])
+                t.add_row([blue(str(sf1)), blue(str(sf2))])
             else:
                 t.add_row([sf1, sf2])
         raise SchemasNotEqualError("\n" + t.get_string())
 
 
 # deprecate this.  ignore_nullable should be a flag.
```
Comment (SemyonSinchenko) [inline chispa/schema_comparer.py:72]: Let's, maybe, add a deprecation warning?

File: chispa/schema_comparer.py [hunk 10]
```diff
@@ -95,11 +103,12 @@
         if not ignore_metadata and sf1.metadata != sf2.metadata:
             return False
         else:
-            return are_datatypes_equal_ignore_nullable(sf1.dataType, sf2.dataType)
+            return are_datatypes_equal_ignore_nullable(sf1.dataType, sf2.dataType)  # type: ignore[no-any-return, no-untyped-call]
 
 
 # deprecate this
```
Comment (SemyonSinchenko) [inline chispa/schema_comparer.py:101]: Let's, maybe, add a deprecation warning?

**Review by SemyonSinchenko — COMMENTED** [review]
I left few comments.

**Comment by SemyonSinchenko** [comment]
@fpgmaas Hi! Is there any chance you can update the PR?

File: chispa/__init__.py (hunk 1 above)
Comment (fpgmaas) [reply chispa/__init__.py:52]: The flag does not seem to be used anywhere.

File: chispa/bcolors.py (hunk 2 above)
Comment (fpgmaas) [reply chispa/bcolors.py:39]: Done

File: chispa/schema_comparer.py (hunk 7 above)
Comment (fpgmaas) [reply chispa/schema_comparer.py:44]: There's multiple functions with `# deprecate this` comment. Let's leave solving that for a follow-up PR.

File: chispa/schema_comparer.py (hunk 8 above)
Comment (fpgmaas) [reply chispa/schema_comparer.py:59]: There's multiple functions with `# deprecate this` comment. Let's leave solving that for a follow-up PR.

File: chispa/schema_comparer.py (hunk 9 above)
Comment (fpgmaas) [reply chispa/schema_comparer.py:72]: There's multiple functions with `# deprecate this` comment. Let's leave solving that for a follow-up PR.

File: chispa/schema_comparer.py (hunk 10 above)
Comment (fpgmaas) [reply chispa/schema_comparer.py:101]: There's multiple functions with `# deprecate this` comment. Let's leave solving that for a follow-up PR.

**Review by SemyonSinchenko — APPROVED** [review]

**Comment by SemyonSinchenko** [comment]
Thank you!

**Pull request merged**